### ISSUE TITLE
How to use this plugin as Pay provider & Apple Pay?

### ISSUE THREAD

**Issue by lorenzOliveto** [flutter-stripe/flutter_stripe#164]
We would like to use this plugin presenting the Apple Pay sheet, obtaining a stripe token, and send it to our backend to complete the payment with stripe.
We've done it with Google Pay, using stripe as the payment gateway in the configuration, but we're not finding a way to do this on iOS & Apple Pay.

I'm assuming this is possible:

> Google Pay: The plugin can easily act as payment provider for the Pay plugin that enables you to seamlessly integrate Google Pay or Apple Pay. All you need to do is add your stripe publishable key to the payment profile.

**Comment by jonasbark** [comment]
Yes, it is possible. Check out this (declined) PR:
https://github.com/google-pay/flutter-plugin/pull/20/files
The only thing for you to do is find out how to extract the actual token from the JSON.

**Comment by lorenzOliveto** [comment]
Thanks for your response, unfortunately I can't make this working, the code I'm using is this:

``` dart
final paymentResult = await payClient.showPaymentSelector(
  provider: PayProvider.apple_pay,
  paymentItems: paymentItems,
);

Map<String, dynamic> tokenMap = json.decode(paymentResult['token']);
String tokenData = tokenMap['data'];

stripe.PaymentMethodParams paymentMethodParams =
    stripe.PaymentMethodParams.cardFromToken(
  token: tokenData,
);
final paymentMethod =
    await stripe.Stripe.instance.createPaymentMethod(paymentMethodParams);
```

The plan is to create the payment method and confirm the payment with the server sending the payment method id (as we're currently doing with the cards).

The createPaymentMethod returns this

`PlatformException(Failed, There was an unexpected error -- try again in a few seconds, null, null)`

**Comment by jonasbark** [comment]
What's in the 'tokenData' variable? A single token or something else?
@jamesblasco can you take a look?

**Comment by lorenzOliveto** [comment]
It's the single token, I've also tried with the whole json string (`paymentResult['token']`) but it gives the same error.

**Comment by niklaesAtMonta** [comment]
@jonasbark Any news regarding this issue?

We are facing the same issue, and the paymentResult coming from Apple contains this (I have deleted some of the contents for privacy reasons):

``` json
{
    "shippingMethod": null, 
    "token": {
        "version": "EC_v1",
        "data": "9YmZQCs5nfsZNo7vFS5BUuaF73Ds+XbHCZ4dnc7oK6GMGBiwT1kYjUpACJ38Bf7ymxzviSYSTJUxLzJiRBCNQlHKDJe+R7hV1Ot4BJwBjxklAIk1w/lpQ0asq2OhPWNDYkT6ca/+iWpH6lJa76fcQDQQXj9qUMb+qNgd8hb+1aPXd90NME3Kn4l2rBCyIjtDlwZZlQaFrjeK2kcPTKmwahaoaG21dikMSguc=",
        "signature": "MIAGCSqGSIb3DQEHAqCAMIACAQExDzANBglghkmMCQGA1UECwwdQXBwbGUgQ2VydGlmaWNhdGlvbiBBdXRob3JpdHkxEzARBgNN1owXzElMCMGA1UEAwwcZWNjLXNtcC1icm9rZXItc2lnbl9VQzQtUFJPRDEUMBIGA1UECwwLaU9TIFN5c3RlbXMxEzARBgNVBAoMCkFwcGxlIEluYy4xCzAJBgNVBAYTAlVTMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEwhV37evWx7Ihj2jdcJChIY3HsL1vLCg9hGCV2Ur0pUEbg0IO2BHzQH6DMx8cVMP36zIg1rrV1O/0komJPnwPE6OCAhEwggINMAwGA1UdEwEB/wQCMAAwHwYDVR0jBBgwFoAUI/JJxE+T5O8n5sT2KGw/orv9LkswRQYIKwYBBQUHAQEEOTA3MDUGCCsGAQUFBzABhilodHRwOi8vb2NzcC5hcHBsZS5jb20vb2NzcDA0LWFwcGxlYWljYTMwMjCCAR0GA1UdIASCARQwggEQMIIBDAYJKoZIhvdjZAUBMIH+MIHDBggrBgEFBQcCAjCBtgyBs1JlbGlhbmNlIG9uIHRoaXMgY2VydGlmaWNhdGUgYnkgYW55IHBhcnR5IGFzc3VtZXMgYWNjZXB0YW5jZSBvZiB0aGUgdGhlbiBhcHBsaWNhYmxlIHN0YW5kYXJkIHRlcm1zIGFuZCBjb25kaXRpb25zIG9mIHVzZSwgY2VydGlmaWNhdGUgcG9saWN5IGFuZCBjZXJ0aWZpY2F0aW9uIHByYWN0aWNlIHN0YXRlbWVudHMuMDYGCCsGAQUFBwIBFipodHRwOi8vd3d3LmFwcGxlLmNvbS9jZXJ0aWZpY2F0ZWF1dGhvcml0eS8wNAYDVR0fBC0wKzApoCegJYYjaHR0cDovL2NybC5hcHBsZS5jb20vYXBwbGVhaWNhMy5jcmwwHQYDVR0OBBYEFJRX22/VdIGGiYl2L35XhQfnm1gkMA4GA1UdDwEB/wQEAwIHgDAPBgkqhkiG92NkBh0EAgUAMAoGCCqGSM49BAMCA0kAMEYCIQC+CVcf5x4ec1tV5a+stMcv60RfMBhSIsclEAK2Hr1vVQIhANGLNQpd1t1usXRgNbEess6Hz6Pmr2y9g4CJDcgs3apjMIIC7jCCAnWgAwIBAgIISW0vvzqY2pcwCgYIKoZIzj0EAwIwZzEbMBkGA1UEAwwSQXBwbGUgUm9vdCBDQSAtIEczMSYwJAYDVQQLDB1BcHBsZSBDZXJ0aWZpY2F0aW9uIEF1dGhvcml0eTETMBEGA1UECgwKQXBwbGUgSW5jLjELMAkGA1UEBhMCVVMwHhcNMTQwNTA2MjM0NjMwWhcNMjkwNTA2MjM0NjMwWjB6MS4wLAYDVQQDDCVBcHBsZSBBcHBsaWNhdGlvbiBJbnRlZ3JhdGlvbiBDQSAtIEczMSYwJAYDVQQLDB1BcHBsZSBDZXJ0aWZpY2F0aW9uIEF1dGhvcml0eTETMBEGA1UECgwKQXBwbGUgSW5jLjELMAkGA1UEBhMCVVMwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAATwFxGEGddkhdUaXiWBB3bogKLv3nuuTeCN/EuT4TNW1WZbNa4i0Jd2DSJOe7oI/XYXzojLdrtmcL7I6CmE/1RFo4H3MIH0MEYGCCsGAQUFBwEBBDowODA2BggrBgEFBQcwAYYqaHR0cDovL29jc3AuYXBwbGUuY29tL29jc3AwNC1hcHBsZXJvb3RjYWczMB0GA1UdDgQWBBQj8knET5Pk7yfmxPYobD+iu/0uSzAPBgNVHRMBAf8EBTADAQH/MB8GA1UdIwQYMBaAFLuw3qFYM4iapIqZ3r6966/ayySrMDcGA1UdHwQwMC4wLKAqoCiGJmh0dHA6Ly9jcmwuYXBwbGUuY29tL2FwcGxlcm9vdGNhZzMuY3JsMA4GA1UdDwEB/wQEAwIBBjAQBgoqhkiG92NkBgIOBAIFADAKBggqhkjOPQQDAgNnADBkAjA6z3KDURaZsYb7NcNWymK/9Bft2Q91TaKOvvGcgV5Ct4n4mPebWZ+Y1UENj53pwv4CMDIt1UQhsKMFd2xd8zg7kGf9F3wsIW2WT8ZyaYISb1T4en0bmcubCYkhYQaZDwmSHQAAMYIBjDCCAYgCAQEwgYYwejEuMCwGA1UEAwwlQXBwbGUgQXBwbGljYXRpb24gSW50ZWdyYXRpb24gQ0EgLSBHMzEmMCQGA1UECwwdQXBwbGUgQ2VydGlmaWNhdGlvbiBBdXRob3JpdHkxEzARBgNVBAoMCkFwcGxlIEluYy4xCzAJBgNVBAYTAlVTAghMMEFJUZ1UNjANBglghkgBZQMEAgEFAKCBlTAYBgkqhkiG9w0BCQMxCwYJKoZIhvcNAQcBMBwGCSqGSIb3DQEJBTEPFw0yMTA3MjMwNTU4MzFaMCoGCSqGSIb3DQEJNDEdMBswDQYJYIZIAWUDBAIBBQChCgYIKoZIzj0EAwIwLwYJKoZIhvcNAQkEMSIEIArJ6rxgEeijW7R0DdcBvTEzhwbO0IQmaCz9GP5/3RpTMAoGCCqGSM49BAMCBEcwRQIgBzwz0IphyZ/xuuhnoZAGnCELZJiJjLMTMDjfMDVKNCICIQDDBGy3uOy6ApgaWKp8vLPjNS95xA1eDuFery1Tu1+dRQAAAAAAAA==",
        "header": {
            "ephemeralPublicKey": "MFkwEwYHKoZIzj0CAQYIKoa/dT3mBB5aBfe5Y0l/womMDcoUW6dBNjepQCo1TWs+tZUBup1SUX4GA==",
            "publicKeyHash": "X2BPdYqQgDdx1zvKx/spy5Q88ihy7jl/H330cL34p/o=",
            "transactionId": "c77d921f456a3e1f8e7fd29e4627f7bf3c8ac78038805943cb9b194a55ac123b"
        }
    }, 
    "billingContact": null, 
    "paymentMethod": 
    {
        "displayName": "Visa 0212", "network": "Visa", "type": 1
    }, 
    "shippingContact": null
}
```

**Comment by jonasbark** [comment]
@jamesblasco mind having a look at this?

**Comment by joknjokn** [comment]
@jamesblasco, if you know of a temporary workaround, please let us know.

**Comment by gospuzic** [comment]
Following this discussion, also required for my payment flow. An update would greatly be appreciated!

**Comment by KevinTheGray** [comment]
Also required for the payment flow in our app, for some context this is possible using the following method: https://github.com/stripe/stripe-ios/blob/21.8.1/Stripe/STPAPIClient%2BApplePay.swift#L18-L26

It's used currently in this deprecated package: 
https://github.com/jonasbark/flutter_stripe_payment/blob/670dc9a539e9a1c9fd480027d830547dc6c1677a/ios/Classes/TPSStripeManager.m#L1242

**Comment by quentinleguennec** [comment]
Also having the same issue, we are trying to save the Apple pay card details for later use (without charging the card yet).

**Comment by jamesblasco** [comment]
Hello everyone,  I am working on an endpoint `createApplePayToken()` that works like:

```dart
final paymentResult = await payClient.showPaymentSelector(provider: provider, paymentItems: items);
final token = await Stripe.instance.createApplePayToken(paymentResult);
print(token.id);
```
If anyone is interested to try with their `merchant_id` in a real device I would really appreciate it

https://github.com/flutter-stripe/flutter_stripe/tree/ios_token

**Comment by jamesblasco** [comment]
This has been published in version 1.3.0 🎉🎉
